File: app/backend/java/payment-service/src/main/java/com/srbms/paymentservice/strategy/PaymentProcessorFactory.java
package com.srbms.paymentservice.strategy;

import com.srbms.paymentservice.entity.PaymentMode;
import com.srbms.paymentservice.exception.PaymentException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
public class PaymentProcessorFactory {

    private final List<PaymentProcessorStrategy> strategies;

    @Autowired
    public PaymentProcessorFactory(List<PaymentProcessorStrategy> strategies) {
        this.strategies = strategies;
    }

    public PaymentProcessorStrategy getProcessor(PaymentMode mode) {
        return strategies.stream()
                .filter(strategy -> strategy.supports(mode))
                .findFirst()
                .orElseThrow(() -> new PaymentException("Unsupported payment mode: " + mode));
    }
}

File: app/backend/.net/SmartRide.PaymentService/Strategies/PaymentProcessorFactory.cs
using SmartRide.PaymentService.Exceptions;
using SmartRide.PaymentService.Models.Enums;

namespace SmartRide.PaymentService.Strategies;

public class PaymentProcessorFactory
{
    private readonly IEnumerable<IPaymentProcessorStrategy> _strategies;

    public PaymentProcessorFactory(IEnumerable<IPaymentProcessorStrategy> strategies)
    {
        _strategies = strategies;
    }

    public IPaymentProcessorStrategy GetProcessor(PaymentMode mode)
    {
        var processor = _strategies.FirstOrDefault(s => s.Supports(mode));
        if (processor == null)
        {
            throw new PaymentException($"Unsupported payment mode: {mode}");
        }
        return processor;
    }
}
